src/charm.py

```diff
@@ -294,35 +296,31 @@ def _reconfigure(self, event) -> None:
         # This code runs on leader_elected event before mongod_pebble_ready
         self._generate_secrets()
 
-        # reconfiguration can be successful only if a replica set is initialised.
         if not self._db_initialised:
             return
 
         with MongoDBConnection(self.mongodb_config) as mongo:
             try:
                 replset_members = mongo.get_replset_members()
-
+                mongodb_hosts = self.mongodb_config.hosts
                 # compare sets of mongod replica set members and juju hosts
                 # to avoid unnecessary reconfiguration.
-                if replset_members == self.mongodb_config.hosts:
+                if replset_members == mongodb_hosts:
+                    self._set_leader_unit_active_if_needed()
                     return
 
                 # remove members first, it is faster
                 logger.info("Reconfigure replica set")
-                for member in replset_members - self.mongodb_config.hosts:
+
+                for member in replset_members - mongodb_hosts:
                     logger.debug("Removing %s from the replica set", member)
                     mongo.remove_replset_member(member)
-                for member in self.mongodb_config.hosts - replset_members:
-                    logger.debug("Adding %s to the replica set", member)
-                    with MongoDBConnection(
-                        self.mongodb_config, member, direct=True
-                    ) as direct_mongo:
-                        if not direct_mongo.is_ready:
-                            logger.debug("Deferring reconfigure: %s is not ready yet.", member)
-                            event.defer()
-                            return
-                    mongo.add_replset_member(member)
 
+                # to avoid potential race conditions -
+                # remove unit before adding new replica set members
+                if type(event) == RelationDepartedEvent and event.unit:
+                    mongodb_hosts = mongodb_hosts - set([self.get_hostname_for_unit(event.unit)])
+                self._remove_unit_from_replica_set(event, mongo, mongodb_hosts - replset_members)
```

Comment:
If we move adding logic into a separate function, maybe we should also move removal logic into a separate function.